### DISCUSSION TITLE
Validate Model

### DISCUSSION THREAD

**Discussion by beppo-ivel** [appcelerator/titanium_mobile#13157] [Q&A] [answered]
If I execute `Alloy.Models.test.set({attr1: "test", attr2: 0})` the validate method is called, contrary the description [here](https://docs.appcelerator.com/platform/latest/#!/guide/Alloy_Backbone_Migration-section-src-43290758_AlloyBackboneMigration-Validation). 

1) How to use `{validate: false}` in this case? I tried `Alloy.Models.test.set({attr1: "test", attr2: 0}, {validate: false})` and `Alloy.Models.test.set({attr1: "test", attr2: 0, validate: false})` which had no effect.
2) Why `Alloy.Models.test.set({attr1: "", attr2: 0})` always returns false if validation went wrong? I expected it would return my custom messages, in that case "title required".

```
validate: function (attrs) {
          for (let key in attrs) {
            let value = attrs[key];

            if(key === "attr1" && value.length < 1){
              Ti.API.error("attr1 test");
              return "attr1 required";
            }

```

**Comment by m1ga** [comment, marked as answer]
Did you check the `Extending the Backbone.Model class` example at https://titaniumsdk.com/guide/Alloy_Framework/Alloy_Guide/Alloy_Models/Alloy_Collection_and_Model_Objects.html#extending-the-backbone-model-class ? That works fine and you can use `isValid()` to check the validation. I don't know the syntax `Alloy.Models.test.set({attr1: "test", attr2: 0})` I use
```
var m = Alloy.createModel("test", {
  attr1:"text1",
  attr2:""
})
``` 
and then m.save() or m.isValid()

**Reply by beppo-ivel** [reply]
> Did you check the Extending the Backbone.Model class example

yes I already checked it before, there I also read out the definition of validate function by extending the model. And that is also the reason why I expected "attr1 required" instead "false" for a invalid attr1 attribute. Otherwise I can't see a reason to return individual messages in the validate-function.

>  I don't know the syntax Alloy.Models.test.set({attr1: "test", attr2: 0}) I use

I use exactly the same syntax. But still further I have to use `set(m)` to change the label text. 

views/index.xml
```
<Alloy>
  <Model src="test" />
    <Window id="window1" title="Window 1" onClose="cleanup">
          <Label id="label1" text="{test.attr1}" />
    </Window>
  </Alloy>
```

controllers/index.js
```
var m = Alloy.createModel("test", {
  attr1:"text1",
  attr2:""
});
Alloy.Models.test.set(m);
```
models/test.js
```
...
extendModel: function(Model) {
      _.extend(Model.prototype, {
        validate: function (attrs) {
          for (let key in attrs) {
            let value = attrs[key];
            if(key === "attr1" && value.length < 1){
              Ti.API.error("attr1 test");
              return "attr1 required";
            }
          }
        },
...
```

**Reply by m1ga** [reply]
ah ok. Please post this code the time you ask your question :) Most of the time my first reply is always to show more code so you should know it.

The answer here should be just increasing the backbone version and use:
`console.log(m.isValid() , m.validationError);` to get your error message.

You can use `"backbone": "1.4.0"` in config.json to use a newer version of Backbone. The `validate:true` part is from the <a href="https://gist.github.com/addyosmani/4629059">Backbone changelog</a> when going from 0.9.2 (default) to a higher version. So if you stay at the default Backbone version it doesn't do anything.

After that you can also write
```
m.set({attr1: "", attr2: 0},{validate:true});
console.log(m.validationError);
```
(see Backbone: https://backbonejs.org/#Model-validationError)

And you'll get your error message.

**Reply by m1ga** [reply]
btw: I've added a PR in order to make the default Backbone version 1.4.0 for new apps: https://github.com/appcelerator/alloy/pull/1243 so it will be less stressful to deal with old version
